game-core/src/main/java/games/strategy/engine/framework/GameRunner.java

```diff
@@ -371,7 +372,7 @@ public static void exitGameIfFinished() {
       final boolean allFramesClosed = Arrays.stream(Frame.getFrames())
           .noneMatch(Component::isVisible);
       if (allFramesClosed) {
-        System.exit(0);
+        ExitStatus.SUCCESS.exit();
```

Comment:
I've been experimenting around with this line, it looks like if we update this to:


```
Arrays.stream(Frame.getFrames())
         .filter(Component::isVisible)
         .filter(Component::dispose);
```
(removing the System.exit), then the app will terminate. The error dialog window that is rendered but hidden is the thing that seems to prevent shutdown.

Reply:
This statement won't do anything, streams are lazy and need a reduction operation at the end in order to do something. (forEach, reduce, collect, *Match, etc.)
So the fact that it works without it indicates that the problem only occured sometimes.

Reply:
Sorry, typo on the second `filter`, should have been a `forEach`
```
Arrays.stream(Frame.getFrames())
         .filter(Component::isVisible)
         .forEach(Component::dispose);
```

By adding that and removing the `System.exit`, I always observed the same behavior either way. In that stream you can add a `peek` to inspect which frames are not getting closed, it's the error dialog. I recall having trouble with that and needing to create it early in case we had an error early. Creating the component was going to be a no-go, but we could toggle it to visible, hence early creation. That component is hidden and not disposed, because of that it's a, if not *the* component that keeps the app open when all other windows are closed.